### ISSUE TITLE
Better diagnostics for over- and under-determined systems using SVD and null space analysis

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#6382]
This is a placeholder created during migration to preserve original issue numbers.

----

Migrated-From: https://trac.openmodelica.org/OpenModelica/ticket/6382

**Comment by casella** [comment]
Following up #14010, @linuslangenkamp implemented a first version of SVD analysis on all sparse Jacobians, activated by `-lv=LOG_NLS_SVD`, see PR #14027. 

It is complemented by the computation of the 1-norm of Jacobian rows and columns, which gives an idea of how good or bad is the scaling of the Jacobian that is passed to the solver. The flag is `-lv=LOG_NLS_JAC_SUMS`, see PR #14083.

**Comment by casella** [comment]
There is one important missing feature for SVD analysis. Currently, in the case of torn systems, the SVD analysis is applied to the reduced systems, only considering tearing variables and residual equations. All information about torn variables and equations is not explicitly available, which makes a detailed analysis basically impossible. The only relevant information obtained is the condtion number

What we need in case of torn equations is to be able to apply the SVD analysis to the _original, un-torn system of equations_, which can be quite large, so it should be implemented by using a sparse SVD algorithm, that only returns the N largest and N smallest singular values only, as well as their corresponding left and right generalized eigenvectors.

Unfortunately, the implementation won't be totally trivial. If the original system is torn, proper start value will be provided for the iteration variables only, whereas all the other variables are computed automatically by executing the first round of assignments of the torn part; then, the obtained unknown vector is used to compute the first Jacobian, which is analyzed via the SVD.

If the original, un-torn equations are analyzed, using the start value for all its unknowns is not a good idea, because most of them would not be OK. There are two cases here:
1. No homotopy used, or lambda = 0 simplified homotopy: in this case one should use the torn equations to compute the full unknown vector, and then use it to compute the full sparse Jacobian for SVD analysis
2. Homotopy is used, lambda > 0 case: in this case one should use the full unknown vector obtained from the previous lambda step of the homotopy transformation

Case 2. is trivial to implement. Case 1. less so, because you need to generate the code both for the torn and for the non-torn equation system. Not sure how to handle that.

**Comment by linuslangenkamp** [comment]
I just restarted work on this ticket. At first, I will add a sparse SVD library to our 3rdParty and allow for sparse SVD of nonlinear systems (with specified #top and #least singular values) similar to the existing LOG_NLS_SVD just sparse - **either torn or un-torn.**

> Case 2. is trivial to implement. Case 1. less so, because you need to generate the code both for the torn and for the non-torn equation system. Not sure how to handle that.

I agree this is definitely not trivial and I will delay this feature for some time.

**Comment by linuslangenkamp** [comment]
I already managed to get the build with [primme](https://github.com/primme/primme) working. Now, only runtime implementations remain

**Comment by casella** [comment]
> I just restarted work on this ticket.

Excellent! We are really looking forward to applying it to some of the failures that have been haunting us for a while.

> At first, I will add a sparse SVD library to our 3rdParty and allow for sparse SVD of nonlinear systems (with specified #top and #least singular values) similar to the existing LOG_NLS_SVD just sparse - **either torn or un-torn.**

OK.
 
> > Case 2. is trivial to implement. Case 1. less so, because you need to generate the code both for the torn and for the non-torn equation system. Not sure how to handle that.
> 
> I agree this is definitely not trivial and I will delay this feature for some time.

I guess for the time being we should try to support Case 1. via `-iif`. This is the scenario I envision:
1. Big nasty torn nonlinear system number XXX fails to converge during initialization, with the supplied start values.
2. The simulation is run again with runtime flag `-saveInitialGuess_system=filename.mat,XXX`. This goes through the initial (or initial-lambda0 in case homotopy is used) system solution, until it reaches system XXX. For that system, it just computes the torn part using the computed initial guesses for the iteration variables, then saves the results obtained so far to the `filename.mat` file.
3. The simulation is run again with runtime flags `-iif=filename.mat -lv_system=XXX -lv=LOG_NLS_SVD`. This loads the previously computed torn variables for all systems up to and including XXX and then shows the SVD log for system XXX. In fact, `-iif` will also load bogus initial guesses for subsequent systems in the BLT, but their log output will be suppressed, so that's no big deal

@linuslangenkamp does this make sense to you? I guess implementing the `-saveInitialGuess_system` flag is child's play compared to generating the code for both torn and untorn systems.

In fact, now that I think of it, you could actually implement a procedure activated by a a single  `-computeInitialGuess_system=XXX` runtime flag (of course to be used in combination with `-lv=LOG_NLS_SVD`) that runs like this:
- Go through the initial (or initial-lambda0 in case homotopy is used) system solution, until system XXX is reached
- For that system, just compute the torn part using the computed initial guesses for the iteration variables
- Saves the results obtained so far to the `init_temp.mat` file
- Re-run the simulation from the start with runtime flags `-iif=init_temp.mat -lv_system=XXX -lv=LOG_NLS_SVD`, exiting after the (attempted or successful) solution of systm XXX and deleting the `init_temp.mat` file.

If this is easy to implement, I would adopt this as the final solution, without bothering about generating code twice. Otherwise I'm fine at using the previously outline procedure with `-saveInitialGuess_system`, we can actually put that in the User Guide, that's no big deal.

**Comment by linuslangenkamp** [comment]
@casella I see no immediate downside in that approach! If we do not have to generate both versions, it is definitely the way to go. Of course we will need some more logic in the runtime but that should be way easier and especially done **faster** compared to the heavy backend work.

I will try to get a first version of the new sparse SVD into the master and then continue work on this
`-computeInitialGuess_system=XXX` flag.

Also, I made good progress on this ticket

- sparse SVD is already working in the runtime
- performed tests on small examples (same solutions as the dense version)

To do:
- output logs via `LOG_NLS_SVD`
- parallelize Jacobian-Vector-Products

The library needs the products $Y := J X_1$ and $Y := J^T X_2$, where $J \in R^{n \times m}$ is the Jacobian, $X_1 \in R^{m \times s}$ and $X_2 \in R^{n \times s}$ are combined block-vectors. In this way the subspace iteration is performed simultanously for all singular values / vectors. We can clearly parallelize this iteration over the columns of $X_1$ and $X_2$, such that the entire procedure scales better. As we link with pthread anyway, I think this is a good idea. I guess we wont stop with 20k x 20k matrices 😉

**Comment by casella** [comment]
> [@casella](https://github.com/casella) I see no immediate downside in that approach!

Good!

> If we do not have to generate both versions it is definitely the way to go. Of course we will need some more logic in the runtime but that should be way easier and especially done **faster** compared to the heavy backend work.

Then, we have a deal.

> I will try to get a first version of the new sparse SVD into the master and then continue work on this `-computeInitialGuess_system=XXX` flag.

Excellent. That would just make do.
 
> Also, I made good progress on this ticket
> 
>     * sparse SVD is already working in the runtime
> 
>     * performed tests on small examples (same solutions as the dense version)

Good!

> The library needs the products Y := J X 1 and Y := J T X 2 , where J ∈ R n × m is the Jacobian, X 1 ∈ R m × s and X 2 ∈ R n × s are combined block-vectors. In this way the subspace iteration is performed simultanously for all singular values / vectors. We can clearly parallelize this iteration over the columns of X 1 and X 2 , such that the entire procedure scales better. As we link with pthread anyway, I think this is a good idea. I guess we wont stop with 20k x 20k matrices 😉

This is an interesting development, in principle. However, I won't spend time on it now (unless it is trivial), for three reasons:
- This SVD calculation is done only once (or a few times, if you want to debug the SVD during the homotopy transformation) at initialization. So, it doesn't have to be super-efficient as the computation of the RHS and Jacobians that may be computed many thousands of times during simulations.
- This feature is used for debugging hard problems. Without it, you may literally spend months trying to figure out what's wrong, possibly without actually getting to a conclusion. So, waiting a couple of minutes, or even a couple of hours, is definitely a non-issue compared to months of frustrating and possibly inconclusive work.
- For sure we won't stop at 20k x 20k, but I hope that given the sparsity, the SVD computation will not scale in a superlinear way. If we can wait a few minutes, we can also wait a few hours. You set up the diagnostic run, go for lunch (or go to bed) and fetch the results of the analysis pointing out where the problem is when you get back to work. Sounds good to me.

Bottom line: a sparse implementation is essential, otherwise we could only handle toy models; a parallel implementation of the sparse computation is probably not necessary, at least in the short-medium term. Maybe in the future, when we can _actually_ run tough, real-life industrial models with 10M equations. Not there yet 😃

**Comment by casella** [comment]
BTW, I just learned that you can use $\LaTeX$ in GitHub tickets. Cool!

**Comment by linuslangenkamp** [comment]
Okay perfect!

> This is an interesting development, in principle. However, I won't spend time on it now (unless it is trivial), for three reasons

Reasonable. I will give it like an hour or so. If it does not work out, its fine as is 😄

> BTW, I just learned that you can use LaTeX in GitHub tickets. Cool!

Yes, its really cool!

**Comment by linuslangenkamp** [comment]
I have something running and was able to actually use the method to make a system converge 😄 

Here is a short demonstrative explanation of the new features:

I started with a poorly conditioned ($\mathrm{cond}(J) > 10^{16}$) Hilbert-like problem
``` modelica
model TestSVD
  parameter Integer n = 12;
  parameter Real xStart[n] = { 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 };
  Real x[n] (start=xStart);
equation
  for j in 1:n loop
    sum((x[i]^2 - i^2) / (1 + i + j) for i in 1:n) = 0;
  end for;
annotation(
    __OpenModelica_simulationFlags(s = "dassl", variableFilter = ".*", lv = "LOG_STDOUT,LOG_ASSERT,LOG_NLS_SVD,LOG_STATS,LOG_SUCCESS", svdCount = "1", nls = "experimental-kinsol"));
end TestSVD;
```
where each variable has start value $1$ and its known that the correct solution is $x_i = \pm i$.

I added the new flag `-svdCount`. This determines how many of the smallest singular values / vectors are calculated. Additionally, we also calculate just the largest singular value such that we get the condition number $\mathrm{cond}(J) > \frac{\sigma_{\max}}{\sigma_{\min}}$. For this example `-svdCount=1`. was fully sufficient! If its set to $0$ or is not set, we perform dense SVD with all singular values. (I think that dense feature is probably obsolete now)

The system failed in the first step, but the diagnostic provided:
```
Smallest right singular vectors (variable space)
    Found 1 singular vectors.
    V[:,12] (singular value 1.38411863e-13)
        V[2][12] = -5.85935248e-01 for NLS Var: 2 with Name: x[6]
        V[6][12] = +4.26359431e-01 for NLS Var: 6 with Name: x[5]
        V[3][12] = +3.62551037e-01 for NLS Var: 3 with Name: x[8]
        V[9][12] = +3.42561882e-01 for NLS Var: 9 with Name: x[11]
...
```
So we know that x[6], x[5] and x[8] are in almost singular direction. Setting their start values to $x_i = i - 0.01$ and performing the next try yields another solver failure, but the diagnostic provides new variables:
```
Smallest right singular vectors (variable space)
    Found 1 singular vectors.
    V[:,12] (singular value 2.79244436e-14)
        V[11][12] = +6.17567681e-01 for NLS Var: 11 with Name: x[7]
        V[9][12] = +5.03118275e-01 for NLS Var: 9 with Name: x[11]
        V[10][12] = -4.08653556e-01 for NLS Var: 10 with Name: x[9]
        V[8][12] = -3.10602649e-01 for NLS Var: 8 with Name: x[10]
....
```
Doing the same with x[7], x[11] and x[9] also failed, but after the next iteration
```
Smallest right singular vectors (variable space)
    Found 1 singular vectors.
    V[:,12] (singular value 7.15356816e-15)
        V[12][12] = +7.79695869e-01 for NLS Var: 12 with Name: x[12]
        V[8][12] = -4.98907452e-01 for NLS Var: 8 with Name: x[10]
        V[11][12] = +2.33324047e-01 for NLS Var: 11 with Name: x[7]
        V[3][12] = -2.27874755e-01 for NLS Var: 3 with Name: x[8]
....
```
and after setting x[12] and x[10] also to $x_i = i - 0.01$, we finally converge!

So we can clearly use this measure to find poor start values.
I did not look into the equation side of things, but they should have similar meaningfulness.

**Comment by linuslangenkamp** [comment]
Now i will need to check if all the stuff (build, build without primme, etc.) is fully functional and doesnt break anything. But I guess we are in place for a PR tomorrow.

**Comment by linuslangenkamp** [comment]
I applied the current routine to the Cammarata initialization problem but it converged very slowly and poorly. I will need to look into the settings as the current state is not feasible to commit

**Comment by casella** [comment]
@linuslangenkamp as far as I understand, the purpose of this diagnostics is not to find better start values to ensure the convergence of solvers; rather, it is to find out which variables/equations are involved in making the system (almost) singular, so that the system itself can be fixed, avoiding the singularity up front.

**Comment by casella** [comment]
> I applied the current routine to the Cammarata initialization problem but it converged very slowly and poorly. I will need to look into the settings as the current state is not feasible to commit

I'm not sure I understand what is converging slowly: the initialization problem or the SVD computation?

**Comment by linuslangenkamp** [comment]
> [@linuslangenkamp](https://github.com/linuslangenkamp) as far as I understand, the purpose of this diagnostics is not to find better start values to ensure the convergence of solvers; rather, it is to find out which variables/equations are involved in making the system (almost) singular, so that the system itself can be fixed, avoiding the singularity up front.

Yes sure! Just wanted to verify that we can find the origin of such poor conditioning.

**Comment by linuslangenkamp** [comment]
> > I applied the current routine to the Cammarata initialization problem but it converged very slowly and poorly. I will need to look into the settings as the current state is not feasible to commit
> 
> I'm not sure I understand what is converging slowly: the initialization problem or the SVD computation?

Sadly the SVD computation took around 100k matrix vector products for the single smallest singular value (matrix is of size ~ 280 x 280). This is very slow and a unreasonable high number of iterations. There **must** be some way of accelerating the algorithm's convergence, e.g. preconditioning or so. I will look into these settings.

**Comment by linuslangenkamp** [comment]
I made some changes and was able to accelerate the convergence by a factor of 20, but the un-torn system (4k x 4k) takes several minutes. There is still something very fishy here, but i cannot find it yet...

**Comment by linuslangenkamp** [comment]
Btw, computing just the largest singular value converges in 127 Matrix-vector ops, taking just 0.01 seconds for the large 4k x 4k un-torn system which is very good. For computing the largest singular value the procedure is pretty much trivial as you can estimate that value by a straightforward Krylov subspace method.

I dont know if my preconditioner is so bad or why the convergence is that slow for the smallest singular values (these are the ones that are of particular interest for the diagnositic).

Edit: @casella Got the runtime for the 4k x 4k system down to some seconds. I will still need to see if its stable enough, but i hope we can soon make a first version, such that you can test if it works on your even larger system.

**Comment by casella** [comment]
> Edit: [@casella](https://github.com/casella) Got the runtime for the 4k x 4k system down to some seconds.

Very good! Anyway, as I wrote previously, a few minutes would still be OK for this kind of analysis, which would still be extremely valuable.

> I will still need to see if its stable enough, but i hope we can soon make a first version, such that you can test if it works on your even larger system.

Question: how do you find a preconditioner which is good enough _in all cases_? With using iterative methods with generic equation-based models, I'm always afraid that they may or may not work well, depending on the structure of the specific model at hand.

Anyway, rest assured we will test this feature in a number of challenging cases. We are really longing for that 😄

**Comment by linuslangenkamp** [comment]
> Question: how do you find a preconditioner which is good enough in all cases? With using iterative methods with generic equation-based models, I'm always afraid that they may or may not work well, depending on the structure of the specific model at hand.

That is exactly what i thought too. For now I am using a simple Jacobi preconditioner which is $M_1^{-1} = \mathrm{diag}(J J^T + \sigma_{\mathrm{est}} I)^{-1}$ or $M_2^{-1} = \mathrm{diag}(J^T J + \sigma_{\mathrm{est}} I)^{-1}$, where $J$ is the Jacobian we want to investigate and $\sigma_{\mathrm{est}}$ is the approximate minimal singular value. I set $\sigma_{\mathrm{est}} = 10^{-10}$ as we assume the matrix is almost singular / poorly conditioned.

There are also more sophisticated methods such as ILU, but I hope that this Jacobi preconditioner is good enough for now.

> Anyway, rest assured we will test this feature in a number of challenging cases. We are really longing for that 😄

Very good, I think its best to not waste too much time on my tests and get something running. Then we will eventually see how well it behaves on the challenging cases.

**Comment by linuslangenkamp** [comment]
Now we get this dump for computing the smallest 4 and largest singular value of the large un-torn 4k x 4k system. As you can see I got the runtime for this system down to about 1.8 seconds which is perfectly fine.

```
// ---------------------------------------------------
#Converged 1 sval[ 0 ]= 1.701572e-08 norm 3.571741e-06 Mvecs 12894 Time 1.73157 stage 1
#Converged 1 sval[ 0 ]= 1.859553e-14 norm 3.436902e-01 Mvecs 12918 Time 1.7371 stage 1
#Converged 2 sval[ 0 ]= 1.644679e-14 norm 3.721267e-01 Mvecs 12980 Time 1.74737 stage 1
#Converged 3 sval[ 2 ]= 2.695695e-14 norm 2.121594e-01 Mvecs 13022 Time 1.75316 stage 1
#Converged 3 sval[ 1 ]= 1.672464e-14 norm 3.588610e-01 Mvecs 13078 Time 1.76563 stage 1
#Converged 4 sval[ 3 ]= 1.584396e-13 norm 4.010108e-02 Mvecs 13936 Time 1.88009 stage 1
LOG_NLS_SVD       | info    | experimental-kinsol: sparse SVD analysis (scaled = true, Caller: experimental-kinsol: Kinsol entry point).
|                 | |       | | Matrix Info
|                 | |       | | | NLS eq index = 5316
|                 | |       | | | Columns      = 3585
|                 | |       | | | Rows         = 3585
|                 | |       | | | NNZ          = 12577
|                 | |       | | | Curr Time    = 0.00000e+00
|                 | |       | | Matrix condition
|                 | |       | | | Cond(M) = 4.29297418e+14
|                 | warning | | | Matrix is very ill-conditioned: 1e12 < Cond(M) = 4.29297418e+14
|                 | info    | | Smallest Singular values
|                 | |       | | | sigma_1   =  7.68005104e-15, rnorm_1   =  7.23383193e-01
|                 | |       | | | sigma_2   =  9.38994576e-15, rnorm_2   =  6.34318749e-01
|                 | |       | | | sigma_3   =  1.87116362e-14, rnorm_3   =  3.01687222e-01
|                 | |       | | | sigma_4   =  1.58439585e-13, rnorm_4   =  3.77764196e-02
|                 | |       | | Largest Singular values
|                 | |       | | | sigma_1   =  3.29702608e+00, rnorm_1   =  2.33755780e-08
|                 | |       | | Smallest right singular vectors (variable space)
|                 | |       | | | Found 4 singular vectors.
|                 | |       | | | V[:,3585] (singular value 7.68005104e-15)
|                 | |       | | | | V[2000][3585] = -6.47393000e-01 for NLS Var: 2000 with Name: stack.module[8].pen.concLosses.Xa_in_Rest[2]
|                 | |       | | | | V[2154][3585] = +6.32812102e-01 for NLS Var: 2154 with Name: stack.module[8].pen.concLosses.Xa_in_Rest[1]
|                 | |       | | | | V[3529][3585] = -3.45226463e-01 for NLS Var: 3529 with Name: stack.module[10].pen.concLosses.Xa_in_Rest[2]
|                 | |       | | | | V[3447][3585] = -1.97815277e-01 for NLS Var: 3447 with Name: stack.module[7].pen.concLosses.Xa_in_Rest[2]
|                 | |       | | | | V[1293][3585] = -8.51514592e-02 for NLS Var: 1293 with Name: stack.module[7].pen.concLosses.Xa_in_Rest[1]
|                 | |       | | | | V[2241][3585] = -7.48070549e-02 for NLS Var: 2241 with Name: stack.module[9].pen.concLosses.Xa_in_Rest[2]
```

I added some new flags to enable better debugging, but we need to see if something is missing e.g. better preconditioners, option to disable the preconditioner, etc.

The new flags are:
- `OMC_LOG_NLS_SVD_V`: increases the internal print level of PRIMME to 2, so we get info about the SVD solving process into stdout (good if you dont know why there is no output / what its doing). It also dumps the setup and algorithms used
- `-svdCount`: As already mentioned this sets the number of smallest singular values / vectors we want to compute (if set to 0 or not set, we perform dense SVD)
- `-svdSigma`: sets the $\sigma_{\mathrm{est}}$ (see above) for the preconditioner, can significantly accelerate the convergence, but should not be taken too small or too large. I set the default to 1e-8, which seems like the optimum for the Cammarata example

**Comment by casella** [comment]
Thank you @linuslangenkamp!

@matteodepascali will now try the SVD analysis on the latest version of his SOFC model. We have evidence of extremely bad conditioning, cond =  $10^{17}$ _on the scaled model_. Can't wait to figure out which equations are the culprit.

**Comment by linuslangenkamp** [comment]
I feel like that I missed the define `OMC_HAVE_PRIMME` for the configure build. I get the message: `Cannot call sparse SVD analysis, because OpenModelica was not build with PRIMME. Set FLAG_SVD_SPARSE_COUNT=0 to perform dense SVD or build OpenModelica with PRIMME via -DOM_OMC_ENABLE_PRIMME=ON.` in the latest nightly or is it working for you? I dont know which build you use for the Windows nightlys.

**Comment by linuslangenkamp** [comment]
> @matteodepascali will now try the SVD analysis on the latest version of his SOFC model. We have evidence of extremely bad conditioning, cond = 10 17 on the scaled model. Can't wait to figure out which equations are the culprit.

Good luck! I hope that you can find the culprit.

**Comment by casella** [comment]
> I feel like that I missed the define `OMC_HAVE_PRIMME` for the configure build. I get the message: `Cannot call sparse SVD analysis, because OpenModelica was not build with PRIMME. Set FLAG_SVD_SPARSE_COUNT=0 to perform dense SVD or build OpenModelica with PRIMME via -DOM_OMC_ENABLE_PRIMME=ON.` in the latest nightly or is it working for you? 

Not sure, we've been busy with other stuff today so we havent' tried yet. @matteodepascali is running these cases on our i9 Linux server, so we actually need the Linux nightly build.

> I dont know which build you use for the Windows nightlys.

For Windows, you can get the git hash of the latest nightly from the filename `here`.

**Comment by linuslangenkamp** [comment]
Since I added a testcase in https://github.com/OpenModelica/OpenModelica/pull/14508 and the tests ran through, we should be good to go. I will try the nightly on monday though.

**Comment by casella** [comment]
OK. @matteodepascali is still working on the proper scaling of the Pyomo code, it turned out there were some equations where the scaling was actually worsened rather than being improved. @linuslangenkamp please let us know when this feature is finalized, so we can test it.

**Comment by linuslangenkamp** [comment]
@casella I made good progress on this ticket and now you can specify
`-saveInitialGuess_system=saved_guess.mat,nls_index` for my kinsol interface `-nls=experimental-kinsol` (will need to unify those soon). It solves all systems before `nls_index` as usual, computes the torn variables in system `nls_index` and writes that initial guess to the specified `saved_guess.mat`.

After that  you can read in the `saved_guess.mat` file with `-iif` and look at system `nls_index` with all the debug flags
`-svdCount`, `-svdSigma` `-lv=LOG_NLS_SVD` / `-lv=LOG_NLS_SVD_V`, `-lv=LOG_NLS_JAC_SUMS`, etc.

Is that already enough or do you need to skip all the other systems when importing that mat file? I guess those are solved anyway, so that should not be an issue right?

> Re-run the simulation from the start with runtime flags -iif=filename.mat -lv_system=XXX -lv=LOG_NLS_SVD, exiting after the (attempted or successful) solution of systm XXX and deleting the init_temp.mat.

Am I missing something else regarding this `-iif`? I do not quite get why you would import the standard `filename.mat` and the `init_temp.mat` (which is the `saved_guess.mat` in my case). Doesnt the `saved_guess.mat` include everything you need, if it was generated by
`-nls=experimental-kinsol -iif=filename.mat -saveInitialGuess_system=saved_guess.mat,nls_index`?

Maybe I should add the feature that the simluation terminates after the specified system has been investigated, then you will not get any further potential non-sense. What do you think?

Besides that, there are just some tests and minor adjustments remaining.

**Comment by linuslangenkamp** [comment]
I tested [PR #14600](https://github.com/OpenModelica/OpenModelica/pull/14600) on two small problems where i generated the torn system, wrote its guess to the file and then imported it for the untorn system. It seems to work fine.

**Comment by casella** [comment]
Thank you @linuslangenkamp, this is really great!

> you can specify `-saveInitialGuess_system=saved_guess.mat,nls_index` for my kinsol interface `-nls=experimental-kinsol` 

OK.

> (will need to unify those soon).

Do you have plans for that? This also involves finalizing the Newton Diagnostics feature

> It solves all systems before `nls_index` as usual, computes the torn variables in system `nls_index` and writes that initial guess to the specified `saved_guess.mat`.

Good.

> After that you can read in the `saved_guess.mat` file with `-iif` and look at system `nls_index` with all the debug flags `-svdCount`, `-svdSigma` `-lv=LOG_NLS_SVD` / `-lv=LOG_NLS_SVD_V`, `-lv=LOG_NLS_JAC_SUMS`, etc.

OK.

> Is that already enough or do you need to skip all the other systems when importing that mat file? I guess those are solved anyway, so that should not be an issue right?

Yeah, don't bother about that.

> > Re-run the simulation from the start with runtime flags -iif=filename.mat -lv_system=XXX -lv=LOG_NLS_SVD, exiting after the (attempted or successful) solution of systm XXX and deleting the init_temp.mat.

So the temp files only works for one attemp, rigth? This seems OK form me, but I need to be sure.

> Am I missing something else regarding this `-iif`? I do not quite get why you would import the standard `filename.mat` and the `init_temp.mat` (which is the `saved_guess.mat` in my case). Doesnt the `saved_guess.mat` include everything you need, if it was generated by `-nls=experimental-kinsol -iif=filename.mat -saveInitialGuess_system=saved_guess.mat,nls_index`?

Yeah, sorry, that was a typo on my side. When I [wrote](https://github.com/OpenModelica/OpenModelica/issues/6382#issuecomment-3385446236)

> Re-run the simulation from the start with runtime flags `-iif=filename.mat -lv_system=XXX -lv=LOG_NLS_SVD`, exiting after the (attempted or successful) solution of systm XXX and deleting the `init_temp.mat`.

I should have written

> Re-run the simulation from the start with runtime flags **`-iif=init_temp.mat`** `-lv_system=XXX -lv=LOG_NLS_SVD`, exiting after the (attempted or successful) solution of systm XXX and deleting the `init_temp.mat` file.

I just edited the [original post](https://github.com/OpenModelica/OpenModelica/issues/6382#issuecomment-3385446236) accordingly.

> Maybe I should add the feature that the simluation terminates after the specified system has been investigated, then you will not get any further potential non-sense. What do you think?

Sounds good. Please do that.
 
> Besides that, there are just some tests and minor adjustments remaining.

Excellent!

FYI, yesterday we managed to get the first system-wide optimization of the @matteodepascali's full SOS-CO2 power plant model (in a simplified version using ideal gas models, for the time being), using the Modelica2Pyomo tool and CONOPT. The resulting NLP has about 1M DAE constraints 😎 . Now those constraints have a condition number of about 1e8, which is reasonable. The reason why we got much worse condition numbers was wrong scaling when derivatives were involved in the Modelica2Pyomo tool, so LOG_NLS_SVD would not have helped much here.

In any case, we'll definitely investigate the conditioning of that model using your feature. cond=1e8 is manageable, but not ideal, something like 1e5 or 1e6 would probably be easier to manage by CONOPT.

**Comment by casella** [comment]
This afternoon I will write a draft documentation of this feature in the User's Guide, including a simple test case, then you can review it.

**Comment by casella** [comment]
We can already start drafting a paper about it for the Asian Modelica Conference next year.

**Comment by linuslangenkamp** [comment]
> Do you have plans for that? This also involves finalizing the Newton Diagnostics feature

I mean we should start it eventually, but I guess a lot of other tickets should be incorporated into such refactoring (e.g. https://github.com/OpenModelica/OpenModelica/issues/14288 or the Jacobian unification Philip and I worked on). We need to prioritize.

>> Is that already enough or do you need to skip all the other systems when importing that mat file? I guess those are solved anyway, so that should not be an issue right?

> Yeah, don't bother about that.

Thought so, very nice.

> So the temp files only works for one attemp, right? This seems OK form me, but I need to be sure.

The temp files `saved_guess.mat` that have been written to disk with `-saveInitialGuess_system=saved_guess.mat,nls_index` are persistent. You can use them several times and they wont be deleted by anything my implementation does. You can specify like `-saveInitialGuess_system=/home/username/documents/savedGuesses/my_new_guess.mat,5133` and it will not be deleted. Just use `-iif=/home/username/documents/savedGuesses/my_new_guess.mat` on other runs and it should work fine. Note that you have to use the `-nls=experimental-kinsol` flag for now, as I only added it for that NLS solver.

>> Maybe I should add the feature that the simluation terminates after the specified system has been investigated, then you will not get any further potential non-sense. What do you think?

> Sounds good. Please do that.

Will add that.

> FYI, yesterday we managed to get the first system-wide optimization of the @matteodepascali's full SOS-CO2 power plant model (in a simplified version using ideal gas models, for the time being), using the Modelica2Pyomo tool and CONOPT. The resulting NLP has about 1M DAE constraints 😎 . 

Amazing! 😎 

> Now those constraints have a condition number of about 1e8, which is reasonable. The reason why we got much worse condition numbers was wrong scaling when derivatives were involved in the Modelica2Pyomo tool, so LOG_NLS_SVD would not have helped much here.

Interesting, so the solution was proper scaling of these derivatives, so something similar to that $1/h$ scaling trick?

> This afternoon I will write a draft documentation of this feature in the User's Guide, including a simple test case, then you can review it.

Seems good!

**Comment by linuslangenkamp** [comment]
Just updated the PR. If it runs through we are ready for merging.
I added the flag -stopAtSystem=nls_index. During the second run, when reading the generated .mat file with the initial guesses, you can specify the nls_index you want to examine. The simulation will then stop immediately after that system has been solved.

**Comment by linuslangenkamp** [comment]
@casella Overall we get something like this:

1. Generate initial guess

* Compile the model with tearing enabled
* Run the model with flags:

```bash
./model_name_torn \
  -nls=experimental-kinsol \
  -saveInitialGuess_system=saved_guess.mat,nls_index
```

* Solves standard initialization up to `nls_index`
* Computes only the torn variables for `nls_index`
* Writes the initial guess to `saved_guess.mat`
* Terminates immediately after `nls_index`


2. Investigate specific system

* Generate untorn model using `--tearingMethod=noTearing` or `minimalTearing` compiler flags
* Run the model with flags:

```bash
./model_name_untorn \
  -iif=saved_guess.mat \
  -nls=experimental-kinsol \
  -stopAtSystem=nls_index \
  -lv=LOG_NLS_SVD \
  -svdCount=3 \
  -anyOtherDebuggingFlag
```

* Loads the initial guess from `saved_guess.mat`
* Solves only specified system `nls_index` (as others should be solved before)
* Use some debugging flags such as `-lv=LOG_NLS_SVD` and `-svdCount=3` to investigate the system
* The simulation is terminated after the system was solved (`-stopAtSystem=nls_index`)
